actors/evm/src/interpreter/precompiles.rs

```diff
@@ -0,0 +1,255 @@
+use std::{borrow::Cow, ops::Mul};
+
+use super::{Message, TransactionAction, H160};
+use fvm_shared::crypto::{
+    hash::SupportedHashes,
+    signature::{SECP_PUB_LEN, SECP_SIG_LEN, SECP_SIG_MESSAGE_HASH_SIZE},
+};
+use substrate_bn::{AffineG1, Fq, Fr, Group, G1};
+
+// TODO it would probably be better to have a deserializer/reader for inputs rather than taking out chunks of arrays
+
+// TODO probably have a different type of input here (probably a deserialized message)
+pub fn is_precompiled(msg: &TransactionAction) -> bool {
+    if let TransactionAction::Call(addr) = msg {
+        !addr.is_zero() && addr <= &MAX_PRECOMPILE
+    } else {
+        false
+    }
+}
+
+// TODO cleanup
+#[derive(Debug)]
+pub struct PrecompileOutput {
+    pub cost: u64,
+    pub output: Vec<u8>,
+}
+
+pub enum ExitReason {
+    Success,
+    OutOfGas,
+    Other(Cow<'static, str>),
+}
+
+impl From<&'static str> for ExitReason {
+    fn from(src: &'static str) -> Self {
+        ExitReason::Other(Cow::Borrowed(src))
+    }
+}
+
+pub type PrecompileFn = fn(&[u8], u64) -> PrecompileResult; // TODO useful error
+pub type PrecompileResult = Result<PrecompileOutput, ExitReason>;
+
+pub fn linear_gas_cost(len: usize, base: u64, word: u64) -> u64 {
+    ((len as u64 + 32 - 1) / 32 * word) + base
+}
+pub fn assert_gas(cost: u64, limit: u64) -> Result<(), ()> {
+    if cost > limit {
+        Err(())
+    } else {
+        Ok(())
+    }
+}
+
+fn nop(inp: &[u8], gas_limit: u64) -> PrecompileResult {
+    todo!()
+}
+
+fn ec_recover(input: &[u8], gas_limit: u64) -> PrecompileResult {
+    let cost = 3_000;
+    let mut buf = [0u8; 128];
+    buf[..input.len().min(128)].copy_from_slice(&input[..input.len().min(128)]);
+
+    let mut hash = [0u8; SECP_SIG_MESSAGE_HASH_SIZE];
+    let mut sig = [0u8; SECP_SIG_LEN];
+
+    hash.copy_from_slice(&input[..32]);
+    sig.copy_from_slice(&input[64..]);
+
+    // recovery byte means a single byte value is 32 bytes long, sad
+    if input[32..63] != [0u8; 31] || !matches!(input[63], 23 | 28) {
+        return Ok(PrecompileOutput { cost, output: Vec::new() });
+    }
+    sig[64] = input[63] - 27;
+
+    let recovered =
+        fvm_sdk::crypto::recover_secp_public_key(&hash, &sig).unwrap_or([0u8; SECP_PUB_LEN]);
```

Comment:
Because 0000… key is an invalid public key so it is equivalent to failure.
The way the ec_recover is the output is compared against known public key.